README
## paleotree: Paleontological and Phylogenetic Analyses of Evolution

paleotree is an R package for transforming, 'a posteriori' time-scaling, and modifying phylogenies containing extinct (i.e. fossil) lineages. In particular, most users are interested in the functions timePaleoPhy, bin_timePaleoPhy, cal3TimePaleoPhy and bin_cal3TimePaleoPhy, which a posteriori time-scale cladograms of fossil taxa into dated phylogenies. This package also contains a large number of likelihood functions for estimating sampling and diversification rates from different types of data available from the fossil record (e.g. range data, occurrence data, etc). paleotree users can also simulate diversification and sampling in the fossil record using the function simFossilRecord, which is a detailed simulator for branching birth-death-sampling processes composed of discrete taxonomic units arranged in ancestor-descendant relationships. Users can use simFossilRecord to simulate diversification in incompletely sampled fossil records, under various models of morphological differentiation (i.e. the various patterns by which morphotaxa originate from one another), and with time-dependent, longevity-dependent and/or diversity-dependent rates of diversification, extinction and sampling. Additional functions allow users to translate simulated ancestor-descendant data from simFossilRecord into standard time-scaled phylogenies or unscaled cladograms that reflect the relationships among taxon units.

The most recent public release of the code is on CRAN at:

	http://cran.r-project.org/web/packages/paleotree/index.html

You can install the most recent public release version of paleotree in R using:

	install.packages("paleotree")

The latest pre-release version of paleotree can be found at github:

	https://github.com/dwbapst/paleotree
	
The most recent commit is currently: [![Build Status](https://travis-ci.org/dwbapst/paleotree.svg?branch=master)](https://travis-ci.org/dwbapst/paleotree) (Travis CI)
	
You can install this latest development version using the R function install_github in the package 'devtools':

	library(devtools)
	install_github("dwbapst/paleotree")

Once installed, you can check the version number of your paleotree install using the R function packageVersion:

	packageVersion("paleotree")

If you use paleotree in your research, you can cite my paper describing paleotree in Methods in Ecology and Evolution:

	Bapst, D.W. 2012. paleotree: an R package for paleontological and phylogenetic analyses of evolution. Methods in Ecology and Evolution. 3: 803-807. doi: 10.1111/j.2041-210X.2012.00223.x
	
	http://onlinelibrary.wiley.com/doi/10.1111/j.2041-210X.2012.00223.x/abstract
	
You can also call the citation for paleotree from within R, using the citation function:
	
	citation("paleotree")
	
This code is authored by David Bapst and offered under CC0.

The current total number of downloads of the paleotree package from the RStudio CRAN mirror is: [![Number of Downloads](http://cranlogs.r-pkg.org/badges/grand-total/paleotree)](https://github.com/metacran/cranlogs.app)

[![Research software impact](http://depsy.org/api/package/cran/paleotree/badge.svg)](http://depsy.org/package/r/paleotree)


CHANGELOG
Version - 3.3.0 - 06-03-19 


Version - 3.2.8 - 06-02-19 (Not Released to CRAN)
- Fixed bug communicated by user Armin Elsler in October 2018 in cal3TimePaleoPhy where if `dateTreatment = 'minMax'` was used, the output trees would have tips at the same absolute time, as the generating time-tree used for cal3 inference was generated using the input fixed first and last dates as if they were FAD and LAD, with the upper age constraint treated as fixed LAD and time of observation for each tip. This has the effect of making 'minMax' treatment with cal3 nearly identical to using 'firstLast'. This is now fixed by creating a new new basic-time-scaled tree using the randomly generated set of point occurrence times within each iteration under 'minMax', such that the resulting tip dates are always different.
- Also fixed tests for variance in cal3 tree calculations that were not properly sorting time data before comparing it against the time-scaled tree.

Version - 3.2.4 - 05-30-19 (Not Released to CRAN)
- Noted that functions for processing MrBayes trees was not checking if names for NEXUS files included .nex or not - added a test for this, so that .nex would be added if it wasn't present
- Anat Shafir reported that paleotree's obtainDatedPosteriorTreesMrB was inappropriately using the prior probability, not the posterior probability, for finding the Maximum A Posterior tree (and thus was presumably returning the Maximum A Priori tree). This was fixed and the help was greatly clarified. For many paleotree users tip-dating with an empty matrix, however, the log-likelihood of the character data is always zero, and thus the posterior probability in those cases was always identical to the prior probability. Regardless, point estimates of topologies are always to be less preferred than looking at samples of trees from the posterior.

Version - 3.2.0 - 04-14-19 (Not Released to CRAN)
-Replaced all instances of is.binary.tree with ape::is.binary.phylo
-Updated all PBDB-dependent functions to use the current 2019 specifications of the v1.2 API for the PBDB
-Phylopic plotting functions refined, with more user control of the resulting plot.
-Numerous fixes to how making taxon trees are created.

Version - 3.1.9 - 02-15-19 (Not Released to CRAN)
-Updates to timePaleoPhy and cal3 documentation for dateTreatment argument, thanks to Armin Elsler for pointing out some inaccuracies in the text.
-Updated some (but not all) functions related to use of PBDB data to use the proper 1.2 API, and made sure usage was properly accounted for.
-Added functions for getting taxonomic data for clades, or for a specific list of taxa from the PBDB, which previously had existed in proto form in examples.
-Added a function for plotting a tree with phylopics from the PBDB - this function currently does not allow for a lot of flexibility and is a work in progress.

Version - 3.1.3 - 10-01-18
-Added function twoWayEcologyCluster for doing two-way R-mode & Q-mode cluster analyses of proportional abundance datasets from community ecology

Version - 3.1.1 - 07-05-18
-Fixed bug with createMrBayesTipDatingNexus to avoid 'morphNexus not found error' when a pre-existing NEXUS file is used but the morphological matrix doesn't need to be entirely reconstituted. Thanks to Struan Henderson for finding this bug.

Version - 3.1.0 - 07-02-18 (Released to CRAN)
-CRAN release
-Modified code checking for unique solutions in minCharChange on advice of Uwe Ligges, to avoid timing-out when tested on openBLAS systems

Version - 3.0.2 - 06-27-18 (Not Released to CRAN)
-Added new function nodeDates2branchLengths, so that node and tip ages can be used to obtain edge lengths, in absolute time, for undated trees, at suggestion of Struan Henderson
-Bumped version number, merged to master branch.

Version - 3.0.1 - 05-03-18 (Not Released to CRAN)
-Added Pete Wagner's functions relating to measuring character exhaustation, added Pete to the authors for paleotree.
-Added another cladogram (an MPT from a 13-taxon analysis) to SongZhangDicrano example dataset
-Added argument verboseWarnings to cal3 functions, modified previous 'warnings' (actually used function for messages) to warnings

Version - 3.0.0 - 10-20-17
-Released to CRAN after a long hiatus from new CRAN releases.

Version - 2.9 - 10-20-17
-Bumped up version number, to reflect ongoing development over a long period without CRAN release. (CRAN release coming soon.)
-Added function createMrBayesTipDatingNexus, which creates a full NEXUS script for doing tip-dating analyses in MrBayes.
-Added function obtainDatedPosteriorTreesMrB, which will get the dated trees from the posterior of a MrBayes analysis (or provide an otherwise difficult to obtain summary tree, appropriate for dated analyses).
-Added function treeContradiction, which gives the contradiction difference (CD) value between two topologies, as described in Bapst et al. in press (Syst Biol).
-Added function setRootAge which will add a $root.time element to a tree, given a set of ages for tip taxa, using the age for the youngest matching taxon name,
-Added function setRootAges for setting root ages for a set of many trees as a multiPhylo object,
-Tried changing changelog.txt to .md, then changed it back.
-So, it turns out that ACCTRAN in phangorn was never converted from Fitch algorithm to Sankoff, and thus it was inappropriate for ACCTRAN to be applied to trees with multifurcations. This now results in an error if attempted by a user. Many apologies!
-Added new dataset RaiaCopesRule, containing dated paleo-phylogenies for ammonite genera, ceratopsian species, and cervid species from Raia et al. 2015, along with ammonite trait data from same publication.
-Fixed error with missing ingroup constraint in createMrBayesTipDatingNexus reported by Melanie Hopkins, thanks!

Version - 2.8 - 04-06-17 (Not Released to CRAN)

R/taxa2phylo.R
#' Convert Simulated Taxon Data into a Phylogeny
#' 
#' Converts temporal and ancestor-descendant relationships of taxa into a
#' time-scaled phylogeny
#' 
#' @details 
#' As described in the documentation for taxa2cladogram, the relationships
#' among morphotaxa in the fossil record are difficult to describe in terms of
#' traditional phylogenies. One possibility is to arbitrarily choose particular
#' instantaneous points of time in the range of some taxa and describe the
#' time-scaled relationships of the populations present at those dates. This is
#' the tactic used by taxa2phylo.
#' 
#' By default, the dates selected ('obs-time' argument) are the last occurences
#' of the taxon, so a simple use of this function will produce a time-scaled
#' tree which describes the relationships of the populations present at the
#' last occurence of each taxon in the sampled data. Alternatively, obs_time
#' can be supplied with different dates within the taxon ranges.
#' 
#' All data relating to when static morpho-taxa appear or disappear in the
#' record is lost; branching points will be the actual time of speciation,
#' which (under budding) will often be in the middle of the temporal range of a
#' taxon.
#' 
#' Cryptic taxa are not dropped or merged as can be done with taxa2cladogram.
#' The purpose of taxa2phylo is to obtain the 'true' pattern of evolution for
#' the observation times, independent of what we might actually be able to
#' recover, for the purpose of comparing in simulation analyses.
#' 
#' As with many functions in the paleotree library, absolute time is always
#' decreasing, i.e. the present day is zero.
#' 
#' @param taxad A five-column matrix of taxonomic data, as output by
#' \code{fossilRecord2fossilTaxa} via simulations produced using \code{simFossilRecord}

#' @param obs_time A vector of per-taxon times of observation which must be in
#' the same order of taxa as in the object taxad; if NULL, the LADs (column 4)
#' in taxad2 are used

#' @param plot Plot the resulting phylogeny?

#' @return The resulting phylogeny with branch lengths is output as an object
#' of class phylo. This function will output trees with the element $root.time,
#' which is the time of the root divergence in absolute time.
#' 
#' The tip labels are the rownames from the simulation input; see documentation
#' for \code{simFossilRecord} and \code{fossilRecord2fossilTaxa} documentation for details.

#' @note DO NOT use this function to time-scale a real tree for a real dataset.
#' It assumes you know the divergence/speciation times of the branching nodes
#' and relationships perfectly, which is almost impossible given the
#' undersampled nature of the fossil record. Use timePaleoPhy or
#' cal3TimePaleoPhy instead.
#' 
#' DO use this function when doing simulations and you want to make a tree of
#' the 'true' history, such as for simulating trait evolution along
#' phylogenetic branches.
#' 
#' Unlike \code{taxa2cladogram}, this function does not merge cryptic taxa in output
#' from \code{simFossilRecord} (via \code{fossilRecord2fossilTaxa})
#' and I do not offer an option to secondarily drop them.
#' The tip labels should provide the necessary information for users to drop
#' such taxa, however. See \link{simFossilRecord}.

#' @author David W. Bapst

#' @seealso \code{\link{simFossilRecord}}, \code{\link{taxa2cladogram}}, \link{fossilRecord2fossilTaxa}

#' @examples
#' 
#' set.seed(444)
#' record <- simFossilRecord(p = 0.1, q = 0.1, nruns = 1,
#'  	nTotalTaxa = c(30,40), nExtant = 0)
#' taxa <- fossilRecord2fossilTaxa(record)
#' #let's use taxa2cladogram to get the 'ideal' cladogram of the taxa
#' tree <- taxa2phylo(taxa)
#' phyloDiv(tree)
#' 
#' #now a phylogeny with tips placed at the apparent time of extinction for each taxon
#' rangesCont <- sampleRanges(taxa,r = 0.5)
#' tree <- taxa2phylo(taxa,obs_time = rangesCont[,2])
#' phyloDiv(tree,drop.ZLB = FALSE)
#' #note that it drops taxa which were never sampled!
#' 
#' #testing with cryptic speciation
#' set.seed(444)
#' record <- simFossilRecord(p = 0.1, q = 0.1, prop.cryptic = 0.5, nruns = 1,
#'  	nTotalTaxa = c(30,40), nExtant = 0, count.cryptic = TRUE)
#' taxaCrypt <- fossilRecord2fossilTaxa(record)
#' treeCrypt <- taxa2phylo(taxaCrypt)
#' layout(1)
#' plot(treeCrypt)
#' axisPhylo()
#' 
#' @export taxa2phylo
taxa2phylo <- function(taxad,obs_time = NULL,plot = FALSE){
		#obs_time = NULL; plot = TRUE
	#INPUT a taxad object and a vector of observation times for each species
		#if obs = NULL, LADs in taxad1 are used as observation times
		#all times must be in backwards format (zero is present)
	#root.time
		#ALL TREES ARE OUTPUT WITH ELEMENTs "$root.time"
		#this is the time of the root on the tree, which is important for comparing across trees
		#this must be calculated prior to adding anything to terminal branches
	#OUTPUT an ape phylo object with the tips at the times of observation
	#require(ape)
	#important checks 06-17-15
	#check that taxa are ordered by FAD
	if(!identical(order(-taxad[,3]),1:nrow(taxad))){
		#let's coerce the taxad so that it satisfies this
		taxadNew <- taxad[order(-taxad[,3]),]
		if(!is.null(obs_time)){
			#reorder obs_time too!
			obs_time <- obs_time[order(-taxad[,3])]
			}
		#reassign ancestor IDs
			#DON'T FORGET YOU NEED TO REORDER THEM
		newAnc <- sapply(taxad[order(-taxad[,3]),2],function(x) {
			if(is.na(x)){
				NA
			}else{
				which(sapply(taxadNew[,1],identical,x))
				}
			})
		if(is.vector(newAnc)){
			taxadNew[,2] <- newAnc
		}else{
			stop("ancestor IDs cannot be reassigned properly")
			}
		#reassign taxon IDs
		taxadNew[,1] <- 1:nrow(taxad)
		taxad <- taxadNew
		#
		#check
		if(!identical(order(-taxad[,3]),1:nrow(taxad))){
			stop("Cannot coerce input table so taxa are ordered with FADs going from first to last")
			}
		#message(paste0("Input table must be ordered with FADs going from first to last",
		#	" \n coercing taxon order to first row"))
		#
		#stop("input table must be ordered with FADs going from first to last")
		}
	# check root ancestor is in first row
	if(!is.na(taxad[1,2])){
		#find the damn root
		isRoot <- which(is.na(taxad[,2]))
		if(length(isRoot) != 1){
			if(length(isRoot)>1){
				stop("Multiple taxa listed as an apparent root (ancestor is NA)")}
			if(length(isRoot)<1){
				stop("No taxa are listed as an apparent root (i.e. ancestor is NA)")}
		}else{
			stop("Root taxon (ancestor listed as NA) must be in row 1")
			#newRoot <- taxad[isRoot,,drop = FALSE]
			#newRoot[,1] <- 1
			#taxadDrop <- taxad[-isRoot,]
			#taxadDrop[taxadDrop[,1] == 1,] <- isRoot
			#taxad <- rbind(newRoot,taxadDrop)
			#taxad <- taxad[order(taxad[,1]),]
			#message(paste0("Root ancestor (ancestor listed as NA) is not in row 1,",
			#	" \n coercing root ancestor to first row"))
			}		
		}
	taxad1 <- taxad[,1:4,drop = FALSE]
	#some checks
	if(!testParentChild(parentChild = taxad1[,2:1])){
		stop("input anc-desc relationships are inconsistent")}
	#
	#convert time so it runs from forward in time
	if(any((taxad1[,4]-taxad1[,3])<0)){
		taxad1[,3:4] <- max(taxad1[,3:4])-taxad1[,3:4]}
	if(any((taxad1[,4]-taxad1[,3])<0)){
		stop("Last occurrences appear to occur before first occurrences?")}
	if(any(table(taxad1[,1])>1)){
		stop(paste("Duplicated values of taxon ID's in input:",
			which(tabulate(taxad1[,1])>1),collapse = " "))}
	if(is.null(obs_time)){
		obs <- taxad1[,4]
	}else{
		obs <- max(taxad[,3:4])-obs_time
		#check if the times of observations are outside of original taxon ranges
		#nameMatch <- match(names(time.obs),rownames(taxR))
		#if(any(is.na(nameMatch))){stop("ERROR: names on time.obs and in candleRes don't match")}
		obsOutRange <- sapply(1:length(obs_time),function(x)
			if(is.na(obs_time[x])){
				FALSE
			}else{
				(obs_time[x]>taxad[x,3])|(obs_time[x]<taxad[x,4])
			})
		if(any(obsOutRange)){
			stop(paste0("ERROR: Given obs_time are outside of the original taxon ranges!",
				"\n If cryptic taxa, perhaps you forgot to set merge.cryptic = FALSE?"))}
		}
	if(nrow(taxad1) != length(obs)){
		stop("Number of observations are not equal to number of lineages!")}
	#make observations as fake taxa, assuming that observations are WITHIN actual taxon ranges
	fake_taxa <- matrix(sapply((1:nrow(taxad1))[!is.na(obs)],function(x)
		 c(nrow(taxad1)+x,taxad1[x,1],obs[x],obs[x])),,4,byrow = TRUE)
	fake_taxa[,1] <- (1:nrow(fake_taxa))+nrow(taxad1)
	taxad2 <- rbind(taxad1,fake_taxa)
	ntaxa <- nrow(taxad2)
	#MAKE IT INTO AN NODE/EDGE-BASED PHYLOGENY
	#find descendents of every taxon
	desc <- lapply(taxad2[,1],function(x)
		taxad2[c(FALSE,taxad2[-1,2] == x),1])	
	#get time of desc births
	births2 <- lapply(desc,function(x)
		 if(length(x)>0){
			sapply(x,function(y) taxad2[y == taxad2[,1],3])
		})
	desc <- lapply(1:length(desc),function(x)	#MORE STUPIDLY COMPLICATED CODE
		if(length(desc[[x]])>0){
			desc[[x]][match(1:length(births2[[x]]),rank(births2[[x]],ties.method = "random"))]
		}else{
			numeric()
			})
	#time of desc births
	births <- lapply(desc,function(x)
		 if(length(x)>0){sapply(x,function(y) taxad2[y == taxad2[,1],3])})	
	#make events list: first event is taxon birth, with that taxon as desc,
		# next is desc births, extinction not recorded but implied
	events <- lapply(1:ntaxa,function(x)
		if(length(desc[[x]])>0){
			c(taxad2[x,1],desc[[x]])
		}else{
			c(taxad2[x,1])
			})
	#times of events: taxon birth, desc births, extinction
	times <- lapply(1:ntaxa,function(x)
		if(length(births[[x]])>0){
			c(taxad2[x,3],births[[x]],taxad2[x,4])
		}else{
			c(taxad2[x,3],taxad2[x,4])
		})
	#labels for each lineage segment between event times
		# (these may as well represent the daughter node ID too)
		#use decimals to keep track of segments, set first segment as X.0
		#as long as a single taxon doesn't have millions of descendants,
			# in which case the IDs may stop being unique...
	nseg <- sapply(times,length)-1
	seg_labs <- lapply(1:ntaxa,function(x) taxad2[x,1]+(1:nseg[x]/(nseg[x]+1)))
	seg_labs <- lapply(seg_labs,function(x) c(floor(x[1]),x[-1]))
	#now, find the mother segment for each taxon
	taxa_anc <- c(0,sapply(2:ntaxa,function(x)
		unlist(seg_labs[taxad2[x,2] == taxad2[,1]])[
			which(unlist(events[taxad2[x,2] == taxad2[,1]]) == taxad2[x,1])-1]))
	#now make list of all seg ids for all mom segs
	moms2 <- lapply(seg_labs,function(x) x[-length(x)])	
	moms <- lapply(1:ntaxa,function(x) c(taxa_anc[x],unlist(moms2[[x]])))
	lengths <- lapply(times,diff)	#get lengths of segments 
	term <- sapply(unlist(seg_labs),function(x) !any(unlist(moms) == x))	#which are terminal?
	#make edge data.frame, with id, anc-id, length and logical indicating terminal branches
	edgeD <- data.frame(id = unlist(seg_labs),anc = unlist(moms),brlen = unlist(lengths),term = term)
	MRCA <- min(edgeD[sapply(edgeD$id,function(x) 1<sum(x == edgeD$anc)),1])
	#edgeD[edgeD[,2] == MRCA,]
	edgeD <- edgeD[-which(edgeD[,1] <= MRCA),]	#want to drop any extraneous root as well
	#add/leave only the terminals needed for the observations, remove others
	droppers <- which(as.logical(edgeD[,4]) & edgeD[,1]<(nrow(taxad1)+1))
	while(length(droppers)>0){
		edgeD <- edgeD[-droppers,]
		edgeD[,4] <- sapply(edgeD[,1],function(x) !any(edgeD[,2] == x))
		droppers <- which(as.logical(edgeD[,4]) & edgeD[,1]<(nrow(taxad1)+1))
		}
	edgeD[,4] <- sapply(edgeD[,1],function(x) !any(edgeD[,2] == x))
	#collapse single internal nodes
	ndesc <- sapply(edgeD[,1],function(x) sum(x == edgeD[,2]))	#ndesc from each node
	while(any(ndesc == 1)){		#picks only internal branches with 1 desc
		#pick a single, if matrix, use first one
		epick <- edgeD[ndesc == 1,]
		if(is.data.frame(epick)){epick <- epick[1,]}	
		edesc <- edgeD[edgeD$anc == epick$id,]
		#remove desc
		edgeD <- edgeD[-which(edgeD$id == edesc$id),]
		#replace picked edge w/new edge
		newe <- data.frame(id = edesc$id, anc = epick$anc,
			brlen = (epick$brlen+edesc$brlen), term = edesc$term)
		edgeD[edgeD$id == epick$id,] <- newe
		ndesc <- sapply(edgeD$id,function(x) sum(x == edgeD$anc))
		}
	ndesc <- sapply(edgeD$id,function(x) sum(x == edgeD$anc))
	#replace stupid decimal edge placeholders with clean numbers
	e_fix <- numeric()
	#the species in the original data
	e_fix[edgeD$term] <- sapply(edgeD[edgeD$term,1],function(x)
		 which(fake_taxa[,1] == x)) 	
	e_fix[!edgeD$term] <- sum(edgeD$term)+1+(1:sum(!edgeD$term))
	ea_fix <- sapply(edgeD$anc,function(x)
		 ifelse(x != MRCA,e_fix[edgeD$id == x],sum(edgeD$term)+1))
	#NOW MAKE A TREE
	tlabs <- rownames(taxad1)[!is.na(obs)]
	edgf <- cbind(ea_fix,e_fix)
	colnames(edgf) <- NULL
	#check one more time before you make it a tree
	if(any(is.na(edgf))){
		stop("NAs introduced into edge matrix?")}
	if(!testParentChild(parentChild = edgf)){
		stop("produced edge matrix is inconsistent")}
	#Now really make it a tree
	tree1 <- list(edge = edgf,tip.label = tlabs,edge.length = edgeD[,3],Nnode = length(unique(edgf[,1])))
	class(tree1) <- "phylo"						#ITS A TREE!
	#tree <- reorder(collapse.singles(tree1),"cladewise") 	#REORDER IT
	#if(!testEdgeMat(tree)){stop("Edge matrix has inconsistencies")}
	#tree <- read.tree(text = write.tree(tree))
	tree <- cleanNewPhylo(tree1)
	#now, root.time should be the time of the first obs PLUS the distance
		# from the earliest tip to the root
	first_obs_time <- max(taxad[,3:4])-min(obs,na.rm = TRUE)
	tree$root.time <- first_obs_time+min(node.depth.edgelength(tree)[1:Ntip(tree)])
	#make it so that root.time-max node dist from root must be below zero
	rootOffset <- tree$root.time-max(node.depth.edgelength(tree))
	if(rootOffset<0){
		tree$root.time <- tree$root.time-rootOffset
		}
	#plot
	if(plot){
		plot(ladderize(tree),show.tip.label = FALSE)
		axisPhylo()
		}
	#
	return(tree)
	}
